IdentityCore/tests/util/cache/MSIDTestCacheFormat.m

```diff
@@ -0,0 +1,295 @@
+    {
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"Missing parameter", nil, nil, nil, nil, nil);
+        }
+        
+        return NO;
+    }
+    
+    NSString *tokenIdentifier = [self tokenIdentifierForAccount:account tokenType:token.tokenType];
+    
+    NSMutableArray *accountTokens = _cacheContents[tokenIdentifier];
+    
+    if (accountTokens)
+    {
+        [accountTokens removeObject:token];
+    }
+    
+    return YES;
+}
+
+#pragma mark - Helpers
+
+- (BOOL)saveTokenForAccount:(MSIDAccount *)account
+                      token:(MSIDToken *)token
+                    context:(id<MSIDRequestContext>)context
+                      error:(NSError **)error
+{
+    if (!token
+        || !account)
+    {
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"Missing parameter", nil, nil, nil, nil, nil);
+        }
+        
+        return NO;
+    }
+    
+    NSString *tokenIdentifier = [self tokenIdentifierForAccount:account tokenType:token.tokenType];
+    
+    @synchronized (self) {
+        
+        NSMutableArray *accountTokens = _cacheContents[tokenIdentifier];
+        
+        if (!accountTokens)
+        {
+            accountTokens = [NSMutableArray array];
+        }
+        
+        [accountTokens addObject:token];
+        _cacheContents[tokenIdentifier] = accountTokens;
+        
+    }
+    
+    return YES;
+}
+
+- (MSIDToken *)getTokenForAccount:(MSIDAccount *)account
+                        tokenType:(MSIDTokenType)tokenType
+                           params:(MSIDRequestParameters *)parameters
+                          context:(id<MSIDRequestContext>)context
+                            error:(NSError * __autoreleasing *)error
+{
+    if (!account
+        || !parameters)
+    {
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"Missing parameter", nil, nil, nil, nil, nil);
+        }
+        
+        return nil;
+    }
+    
+    NSString *tokenIdentifier = [self tokenIdentifierForAccount:account tokenType:tokenType];
+    
+    NSMutableArray *accountTokens = _cacheContents[tokenIdentifier];
+    
+    if (!accountTokens || ![accountTokens count])
+    {
+        return nil;
+    }
+    
+    return accountTokens[0];
+}
+
+- (NSString *)tokenIdentifierForAccount:(MSIDAccount *)account
+                              tokenType:(MSIDTokenType)tokenType
+{
+    NSString *userIdentifier = account.userIdentifier;
+    
+    if (!userIdentifier)
+    {
+        userIdentifier = account.upn;
+    }
+    
+    return [NSString stringWithFormat:@"%@_%@", userIdentifier, [self tokenTypeAsString:tokenType]];
+}
+
+- (NSString *)tokenTypeAsString:(MSIDTokenType)tokenType
+{
+    NSString *typeIdentifier = @"at";
+    
+    if (tokenType == MSIDTokenTypeRefreshToken)
+    {
+        typeIdentifier = @"rt";
+    }
+    
+    return typeIdentifier;
```

Comment:
how about a ADFS token?

Reply:
We treat it as Access token in the internal implementation to mimic the actual implementation.